pkg/fqdn/service/service.go

```diff
@@ -102,17 +123,112 @@ type PolicyUpdater interface {
+//  3. Send the current state of the DNS rules to the client (We store the current state of
+//     DNS rules during the endpoint regeneration see UpdatePolicyRules)
+//  4. Wait for the context to be done and return the error if any
+//  5. Delete the stream from the local map
 func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
-	// This is a temporary implementation to send the current state of the DNS rules to the client and used for testing
-	stream.Send(&pb.PolicyState{RequestId: "test"})
-	return nil
+	streamCtx, cancel := context.WithCancel(stream.Context())
+	s.streams.Store(stream, cancel)
+
+	ackRecvStream := make(chan struct{}, 1)
+	errChan := make(chan error)
+	// Start a goroutine to receive the DNS policies ACKs
+	go func() {
+		if err := s.receiveDNSPolicyACKs(stream, ackRecvStream); err != nil {
+			s.log.Error("Error receiving DNS policies ACK", logfields.Error, err)
+			cancel() // Cancel the context to close the stream
+			s.deleteStream(stream)
+			errChan <- err
+			return
+		}
+	}()
+	// Wait for the ACK channel to be closed
+	select {
+	case <-ackRecvStream:
+		// ACK channel is closed, we can proceed
+	case err := <-errChan:
+		// Error receiving the ACK
+		if err != nil {
+			s.log.Error("Error receiving DNS policies ACK", logfields.Error, err)
+			return err
+		}
+	}
+	s.log.Debug("Received ACK for DNS policies")
+
+	// Send the current state of the DNS policies to the client
+	s.log.Debug("Sending current state of DNS policies to the client")
+	if err := s.UpdatePolicyRules(nil, false); err != nil {
+		s.log.Error("Error sending current state of DNS rules", logfields.Error, err)
+		cancel() // Cancel the context to close the stream
+		s.deleteStream(stream)
+		return err
+	}
+
+	s.log.Debug("Waiting for context to be done for streaming DNS policies")
+	<-streamCtx.Done()
+	err := streamCtx.Err()
+	s.log.Warn("Context done for streaming DNS policies", logfields.Error, err)
+	s.deleteStream(stream)
+	return err
+}
+
+// receiveDNSPolicyACKs receives the DNS policies ACKs from the client
+// If the success is false, we can send cancel signal to the channel
+// in that case SDP will recreate the stream.
+// The flow of the method is as follows:
+// 1. Close the ACK channel for the stream to send the current state of the DNS policies
+// 2. Start a loop to receive the DNS policies ACKs
+// 3. If the context is done, return the error
+// 4. If the error is not nil, return the error
+// 5. If the response code is not NO_ERROR, send cancel signal to the channel
+// 6. Delete the request from the map
+func (s *FQDNDataServer) receiveDNSPolicyACKs(stream pb.FQDNData_StreamPolicyStateServer, ack chan struct{}) error {
+	var once sync.Once
+
+	for {
+		select {
+		case <-stream.Context().Done():
+			s.log.Info("Stream context is finished, closing the stream")
+			return stream.Context().Err()
+		default:
+			// Continue receiving the DNS policies ACKs
+		}
+		once.Do(func() {
```

Comment:
Shouldn't this be after the `Recv()`?

Reply:
With each invocation to connect to cilium agent, we spawn a new go routine to handle the receiving of ack from sdp and then send the rules to that sdp. 
when you say three go routines, are you referring to the call to `StreamPolictState` as well ? 
(In my head, I was just creating one new go routine per connection for receiving acks. )